src/qililab/qprogram/quantum_machines_compiler.py

```diff
@@ -163,36 +166,40 @@ def traverse(block: Block):
         return qua_program, self._configuration, measurements
 
     def _process_measurements(self):
-        measurements = []
-
-        for i, measurement_compilation_info in enumerate(self._measurements):
-            measurement = MeasurementInfo()
-            if (stream_I := measurement_compilation_info.stream_I) is not None:
-                for loop_iteration in measurement_compilation_info.loops_iterations:
+        def _process_measurement(measurement: _MeasurementCompilationInfo, index: int):
+            result_handles: list[str] = []
+            if measurement.stream_I is not None:
+                stream_I = measurement.stream_I
+                for loop_iteration in measurement.loops_iterations:
                     stream_I = stream_I.buffer(loop_iteration)
-                if measurement_compilation_info.average:
+                if measurement.average:
                     stream_I = stream_I.average()
-                stream_I.save(f"I_{i}")
-                measurement.result_handles.append(f"I_{i}")
-            if (stream_Q := measurement_compilation_info.stream_Q) is not None:
-                for loop_iteration in measurement_compilation_info.loops_iterations:
+                stream_I.save(f"I_{index}")
+                result_handles.append(f"I_{index}")
+            if measurement.stream_Q is not None:
+                stream_Q = measurement.stream_Q
+                for loop_iteration in measurement.loops_iterations:
                     stream_Q = stream_Q.buffer(loop_iteration)
-                if measurement_compilation_info.average:
+                if measurement.average:
                     stream_Q = stream_Q.average()
-                stream_Q.save(f"Q_{i}")
-                measurement.result_handles.append(f"Q_{i}")
-            if measurement_compilation_info.stream_raw_adc is not None:
-                input1 = measurement_compilation_info.stream_raw_adc.input1()
-                input2 = measurement_compilation_info.stream_raw_adc.input2()
-                for loop_iteration in measurement_compilation_info.loops_iterations:
-                    input1 = input1.buffer(loop_iteration)
-                    input2 = input2.buffer(loop_iteration)
-                input2.save_all(f"adc1_{i}")
-                input2.save_all(f"adc2_{i}")
-                measurement.result_handles.append(f"adc1_{i}")
-                measurement.result_handles.append(f"adc2_{i}")
-
-            measurements.append(measurement)
+                stream_Q.save(f"Q_{index}")
+                result_handles.append(f"Q_{index}")
+            if measurement.stream_raw_adc is not None:
```

Comment:
Just a question, I guess this means both adc1 and adc2 always are present together? If one exists so there is the other?

Reply:
Nice catch. I though so, and it is true for the drive buses, but I looked more carefully in the examples and it actually depends on if the element has one or two outputs defined.

```
"rr1": "outputs": {"out1": ("con1", 2)},
```

vs

```
"rr1": "outputs": {"out1": ("con1", 1), "out2": ("con1", 2)},
```

But this information is currently not know to the compiler (since it depends on the runcard provided).

Reply:
I refactored this part, so `input1` and `input2` are processed separately, and added a `TODO` to make the check. (it's better to implement this after having the translation from runcard to dict, so we figure out what is the best way to pass the information to the compiler)

Reply:
I agree with the refactor and then check once we have the translation.